crates/namada/src/ledger/protocol/mod.rs

```diff
@@ -600,18 +652,130 @@ where
                 "Transfer of tx fee cannot be applied to due to fee overflow. \
                  This shouldn't happen."
             );
+            shell_params.state.write_log_mut().drop_batch();
 
             Err(Error::FeeError(format!("{}", e)))
         }
     }
 }
 
-/// Transfer `token` from `src` to `dest`. Returns an `Err` if `src` has
-/// insufficient balance or if the transfer the `dest` would overflow (This can
-/// only happen if the total supply doesn't fit in `token::Amount`). Contrary to
-/// `crate::token::transfer` this function updates the tx write log and
-/// not the block write log.
-fn token_transfer<WLS>(
+fn try_masp_fee_payment<S, D, H, CA>(
+    ShellParams {
+        tx_gas_meter,
+        state,
+        vp_wasm_cache,
+        tx_wasm_cache,
+    }: &mut ShellParams<'_, S, D, H, CA>,
+    tx: &Tx,
+    tx_index: &TxIndex,
+) -> Result<Option<(BatchedTxResult, Hash)>>
+where
+    S: State<D = D, H = H>
+        + StorageRead
+        + Read<Err = namada_state::Error>
+        + Sync,
+    D: 'static + DB + for<'iter> DBIter<'iter> + Sync,
+    H: 'static + StorageHasher + Sync,
+    CA: 'static + WasmCacheAccess + Sync,
+{
+    // The fee payment is subject to a gas limit imposed by a protocol
+    // parameter. Here we instantiate a custom gas meter for this step and
+    // initialize it with the already consumed gas. The gas limit should
+    // actually be the lowest between the protocol parameter and the actual gas
+    // limit of the transaction
+    let min_gas_limit = state
+        .read::<u64>(
+            &namada_parameters::storage::get_masp_fee_payment_gas_limit_key(),
+        )
+        .expect("Error reading the storage")
+        .expect("Missing masp fee payment gas limit in storage")
+        .min(tx_gas_meter.borrow().tx_gas_limit.into());
+
+    let mut gas_meter = TxGasMeter::new(
+        namada_gas::Gas::from_whole_units(min_gas_limit).ok_or_else(|| {
+            Error::GasError("Overflow in gas expansion".to_string())
+        })?,
+    );
+    gas_meter
+        .copy_consumed_gas_from(&tx_gas_meter.borrow())
+        .map_err(|e| Error::GasError(e.to_string()))?;
+    let ref_unshield_gas_meter = RefCell::new(gas_meter);
+
+    let valid_batched_tx_result = {
+        // NOTE: A clean tx write log must be provided to this call
+        // for a correct vp validation. Block and batch write logs, instead,
+        // should contain any prior changes (if any). This is to simulate
```

Comment:
We had an issue with the old implementation of masp fee payment (fee unshielding) for which some keys (the tx hash if I remember right) were in the write log when we launched the wasm execution leading to a vp rejection. I'll check if this is still the case and if not I'll remove the precommit field of the write log

Reply:
I don't quite follow - that sounds like a potential bug?

Reply:
So the tx is indeed the first in the batch but before executing it we might write some keys in the write log (at the moment I believe it's only the hash of the tx for replay protection). This was (in the old implementation) leading to one of the vps (the masp one if I recall correctly) rejecting the transaction. I believe this is no more the case because:

1. We write the tx hash directly in the block write log and so this should not appear in the set of changed keys that we pass to the vp
2. The masp vp (and also the other ones) should now allow for foreign keys by simply ignoring them (it wasn't the case before I think)

So we might be able to remove this logic from here, but I'm still checking

Reply:
Removed in 9e4d4c9693499bf366841db3efb663f81b247c16